ouroboros-network/test/Test/Ouroboros/Network/KeepAlive.hs

```diff
@@ -136,6 +137,67 @@ instance Arbitrary NetworkDelay where
         m <- choose (1, 1000 :: Int) -- A delay between 1 and 1000 ms
         return $ NetworkDelay $ (fromIntegral m) / 1000
 
+prop_keepAlive_ping_pongM
+    :: forall m.
+        ( Eq (Async m ())
+        , MonadAsync m
+        , MonadFork m
+        , MonadMask m
+        , MonadMonotonicTime m
+        , MonadSay m
+        , MonadST m
+        , MonadSTM m
+        , MonadTimer m
+        , MonadThrow (STM m)
+        )
+    => NetworkDelay
+    -> Int
+    -> m Property
+prop_keepAlive_ping_pongM (NetworkDelay nd) seed = do
+    registry <- newFetchClientRegistry
+    controlMessageV <- newTVarM Continue
+    let controlMessageSTM = readTVar controlMessageV
+        clientId = "client"
+        keepAliveInterval = 10
+
+    (c_aid, s_aid) <- runKeepAliveClientAndServer (NetworkDelay nd) seed verboseTracer controlMessageSTM
+                          registry clientId (KeepAliveInterval keepAliveInterval)
+    threadDelay 15 -- enough time to send one, and only one keep-alive
+    dqLive <- atomically $ readPeerGSVs registry
+
+    atomically $ writeTVar controlMessageV Terminate
+    void $ wait c_aid
+    void $ wait s_aid
+
+    -- XXX Must be larger than the KeepAliveInterval timeout or we leak threads in the SIM
```

Comment:
I have an impression that this would help: https://github.com/input-output-hk/ouroboros-network/pull/2631 (`registerDelay` without unsupervised threads). It might be worth trying.

Reply:
If the PR gets merged before this one, I'll rebase and try it.